src/lib/market-service/portals/portals.ts

```diff
@@ -0,0 +1,309 @@
+import type { ChainId } from '@shapeshiftoss/caip'
+import {
+  arbitrumChainId,
+  ASSET_NAMESPACE,
+  avalancheChainId,
+  baseChainId,
+  bscChainId,
+  ethChainId,
+  fromAssetId,
+  gnosisChainId,
+  optimismChainId,
+  polygonChainId,
+  toAssetId,
+} from '@shapeshiftoss/caip'
+import type {
+  AssetsByIdPartial,
+  FindAllMarketArgs,
+  HistoryData,
+  MarketCapResult,
+  MarketData,
+  MarketDataArgs,
+  PriceHistoryArgs,
+} from '@shapeshiftoss/types'
+import { HistoryTimeframe } from '@shapeshiftoss/types'
+import Axios from 'axios'
+import { setupCache } from 'axios-cache-interceptor'
+import dayjs from 'dayjs'
+import qs from 'qs'
+import { zeroAddress } from 'viem'
+import { bn, bnOrZero } from 'lib/bignumber/bignumber'
+import { assertUnreachable, getTimeFrameBounds, isToken } from 'lib/utils'
+
+import generatedAssetData from '../../asset-service/service/generatedAssetData.json'
+import type { MarketService } from '../api'
+import { DEFAULT_CACHE_TTL_MS } from '../config'
+import { isValidDate } from '../utils/isValidDate'
+import type { GetTokensResponse, HistoryResponse } from './types'
+
+const calculatePercentChange = (openPrice: string, closePrice: string): number => {
+  const open = bnOrZero(openPrice)
+  const close = bnOrZero(closePrice)
+  if (open.isZero()) return 0
+  return close.minus(open).dividedBy(open).times(100).toNumber()
+}
+
+export const CHAIN_ID_TO_PORTALS_NETWORK: Partial<Record<ChainId, string>> = {
+  [avalancheChainId]: 'avalanche',
+  [ethChainId]: 'ethereum',
+  [polygonChainId]: 'polygon',
+  [bscChainId]: 'bsc',
+  [optimismChainId]: 'optimism',
+  [arbitrumChainId]: 'arbitrum',
+  [gnosisChainId]: 'gnosis',
+  [baseChainId]: 'base',
+}
+
+const { throttle, clear } = (({
+  capacity,
+  costPerReq,
+  drainPerInterval,
+  intervalMs,
+}: {
+  capacity: number
+  costPerReq: number
+  drainPerInterval: number
+  intervalMs: number
+}) => {
+  let currentLevel = 0
+  let pendingResolves: ((value?: unknown) => void)[] = []
+
+  const drain = () => {
+    const drainAmount = Math.min(currentLevel, drainPerInterval)
+    currentLevel -= drainAmount
+
+    // Resolve pending promises if there's enough capacity
+    while (pendingResolves.length > 0 && currentLevel + costPerReq <= capacity) {
+      const resolve = pendingResolves.shift()
+      if (resolve) {
+        currentLevel += costPerReq
+        resolve()
+      }
+    }
+  }
+
+  // Start the interval to drain the capacity
+  const intervalId = setInterval(drain, intervalMs)
+
+  const throttle = async () => {
+    if (currentLevel + costPerReq <= capacity) {
+      // If adding another request doesn't exceed capacity, proceed immediately
+      currentLevel += costPerReq
+    } else {
+      // Otherwise, wait until there's enough capacity
+      await new Promise(resolve => {
+        pendingResolves.push(resolve)
+      })
+    }
+  }
+
+  const clear = () => clearInterval(intervalId)
+
+  return { throttle, clear }
+})({
```

Comment:
Note, this guy (`createThrottle` as a singleton) has been redeclared because we can't import from `scripts` - didn't feel like fiddling with tsconfig nor adding a root utils directory as in the scope of this PR

Reply:
I came here to ask why we weren't re-using `createThrottle` - I found my answer 🙏